sw/device/tests/rv_core_ibex_address_translation_test.c

```diff
@@ -0,0 +1,295 @@
+  // restore control flow to the `test_main` function as described
+  // above.
+  uintptr_t ret_addr = *(uintptr_t *)(mepc_stack_addr + OTTF_WORD_SIZE);
+
+  uint32_t mcause = ibex_mcause_read();
+  ottf_exc_id_t exception_id = mcause & kIdMax;
+
+  switch (exception_id) {
+    case kInstrAccessFault:
+      LOG_INFO("Instruction access fault handler");
+      access_fault = true;
+      *(uintptr_t *)mepc_stack_addr = ret_addr;
+      break;
+    default:
+      LOG_FATAL("Unexpected exception id = 0x%x", exception_id);
+      abort();
+  }
+}
+
+/**
+ * Takes a unsigned 32bit integer and rounds it to the next power of 2.
+ * Algorithm explained here:
+ *   https://graphics.stanford.edu/~seander/bithacks.html#RoundUpPowerOf2
+ *
+ * @param n The number to round.
+ * @return The next power of two from the given number.
+ */
+uint32_t next_power_of_two(uint32_t n) {
+  n--;
+  n |= n >> 1;
+  n |= n >> 2;
+  n |= n >> 4;
+  n |= n >> 8;
+  n |= n >> 16;
+  return ++n;
+}
+
+/**
+ * Sets up the flash controller.
+ *
+ * @return The flash controller's handle.
+ */
+dif_flash_ctrl_state_t init_flash() {
+  dif_flash_ctrl_state_t flash_ctrl;
+
+  CHECK_DIF_OK(dif_flash_ctrl_init_state(
+      &flash_ctrl,
+      mmio_region_from_addr(TOP_EARLGREY_FLASH_CTRL_CORE_BASE_ADDR)));
+  flash_ctrl_testutils_wait_for_init(&flash_ctrl);
+
+  // Set up default access for data partitions.
+  flash_ctrl_testutils_default_region_access(
+      &flash_ctrl, /*rd_en=*/true, /*prog_en=*/true, /*erase_en=*/true,
+      /*scramble_en=*/false, /*ecc_en=*/false, /*high_endurance_en=*/false);
+
+  return flash_ctrl;
+}
+
+/**
+ * Configures the given address translation mapping in the given slot
+ * for both IBus and DBus.
+ *
+ * @param ibex_core A handle to the ibex core.
+ * @param slot The slot to be used for mapping.
+ * @param mapping A description of the mapping.
+ */
+void map_to_slot(dif_rv_core_ibex_t *ibex_core,
+                 dif_rv_core_ibex_addr_translation_slot_t slot,
+                 dif_rv_core_ibex_addr_translation_mapping_t mapping) {
+  for (size_t idx = 0; idx < 2; ++idx) {
+    CHECK_DIF_OK(dif_rv_core_ibex_configure_addr_translation(
+        ibex_core, slot, kBuses[idx], mapping));
+  };
+}
+
+/**
+ * Enables IBus and DBus address translation for the given slot.
+ *
+ * @param ibex_core A handle to the ibex core.
+ * @param slot The slot to be used enabled.
+ */
+void enable_slot(dif_rv_core_ibex_t *ibex_core,
+                 dif_rv_core_ibex_addr_translation_slot_t slot) {
+  for (size_t idx = 0; idx < 2; ++idx) {
+    CHECK_DIF_OK(
+        dif_rv_core_ibex_enable_addr_translation(ibex_core, slot, kBuses[idx]));
+  };
+}
+
+bool test_main(void) {
+  // Calculate 2-byte aligned addresses to put a copy of the toy functions in.
+  uintptr_t flash_mem_end_addr = TOP_EARLGREY_FLASH_CTRL_MEM_BASE_ADDR +
+                                 TOP_EARLGREY_FLASH_CTRL_MEM_SIZE_BYTES;
+  uintptr_t offset = 0x1000;
```

Comment:
Should you check whether this offset is greater than `kMakeLowerCaseFnSize` and `kGetNameFnSize`?